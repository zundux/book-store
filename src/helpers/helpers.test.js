import { findFirstKey, flattenObj, focusOnFirstInvalidField, getFetchParams } from "./index";

describe("helper functions", () => {

  it("get flatten object from an object or an array", () => {
    const object = {
      a: 1,
      b: {
        bb: "text",
        bbb: {
          bbbb: [1, 2, 3, 4],
          bbbbb: [
            {
              id: 11
            },
            {
              id: 12,
              authors: ["author 1", "author 2"]
            }
          ]
        }
      }
    };
    const array = [object, 123];

    expect(flattenObj(object)).toMatchSnapshot();
    expect(flattenObj(array)).toMatchSnapshot();
  });

  it("get first by value", () => {
    const errors = {
      "first-name": undefined,
      "last-name": "required",
      "email": undefined,
      "phone": "not valid",
      "address": undefined
    };

    expect(findFirstKey(value => value !== undefined)(errors)).toMatchSnapshot();
  });

  it("call focus on form field (scroll window to field and set the field active)", () => {
    const firstInvalidFieldName = "last-name";
    const errors = {
      "first-name": undefined,
      [firstInvalidFieldName]: "required",
      "email": undefined,
      "phone": "not valid",
      "address": undefined
    };

    const focus = jest.fn();
    const getBoundingClientRect = jest.fn(() => {
      return { top: 300 };
    });
    const element = {
      focus,
      getBoundingClientRect
    };
    const scrollTo = jest.fn();
    const getElementsByName = jest.fn(() => [element]);
    document.getElementsByName = getElementsByName;
    window.scrollTo = scrollTo;
    window.pageYOffset = 150;

    focusOnFirstInvalidField(errors);

    expect(getElementsByName).toBeCalledWith(firstInvalidFieldName);
    expect(getBoundingClientRect).toHaveBeenCalled();
    expect(focus).toHaveBeenCalled();
    expect(scrollTo).toBeCalledWith(0, 350);
  });


  it("create a fetching parameters object", () => {
    const defaultParams = {
      q: "",
      filter: "full"
    };
    const params = { q: "man in black" };

    const expectedParams = {
      q: "man in black",
      filter: "full"
    };
    expect(getFetchParams(defaultParams, params)).toEqual(expectedParams);
  });
});