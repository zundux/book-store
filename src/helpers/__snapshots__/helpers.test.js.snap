// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`helper functions get first by value 1`] = `"last-name"`;

exports[`helper functions get flatten object from an object or an array 1`] = `
Object {
  "a": 1,
  "b.bb": "text",
  "b.bbb.bbbb.0": 1,
  "b.bbb.bbbb.1": 2,
  "b.bbb.bbbb.2": 3,
  "b.bbb.bbbb.3": 4,
  "b.bbb.bbbbb.0.id": 11,
  "b.bbb.bbbbb.1.authors.0": "author 1",
  "b.bbb.bbbbb.1.authors.1": "author 2",
  "b.bbb.bbbbb.1.id": 12,
}
`;

exports[`helper functions get flatten object from an object or an array 2`] = `
Object {
  "0.a": 1,
  "0.b.bb": "text",
  "0.b.bbb.bbbb.0": 1,
  "0.b.bbb.bbbb.1": 2,
  "0.b.bbb.bbbb.2": 3,
  "0.b.bbb.bbbb.3": 4,
  "0.b.bbb.bbbbb.0.id": 11,
  "0.b.bbb.bbbbb.1.authors.0": "author 1",
  "0.b.bbb.bbbbb.1.authors.1": "author 2",
  "0.b.bbb.bbbbb.1.id": 12,
  "1": 123,
}
`;
